### ISSUE TITLE
Invalid memory access when using parallel_for and level 1 scratch memory.

### ISSUE THREAD

**Issue by jfields7** [kokkos/kokkos#6398]
**Describe the bug**
I recently updated to Kokkos 4.1.00 (1a3ea28), and since then I've started running into issues with invalid memory accesses for `Kokkos::parallel_for` loops. If I downgrade to 4.0.00 (aa1f48f), the problem disappears, so it appears to be due to a recent change (or bug?) in the code.

It seems that a `parallel_for` loop of sufficient size with level 1 scratch memory on CUDA will eventually produce an error similar to the following:
`cudaFreeHost(constantMemHostStaging) error( cudaErrorIllegalAddress): an illegal memory access was encountered /home/jacob/projects/experiments/scratch/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:589`
In my case, the minimum league size producing the error is 7535, though I have no reason to believe that is some magic number fixed across all problem sizes, configurations, and machines.

**Please also include the following items to support reproducing the bug**
1. compiler: nvcc 11.7.64
2. Kokkos release or commit used (i.e., the sha1 number): Kokkos 4.1.00 (1a3ea28)
3. platform and backend: WSL2 Ubuntu 20.04, CUDA backend
4. cmake configure command: `cmake -DKokkos_ENABLE_CUDA=ON -DKokkos_ARCH_AMPERE86=ON -DKokkos_ARCH_ZEN2=ON ../`
5. output from cmake command:
```
-- The C compiler identification is GNU 9.4.0
-- The CXX compiler identification is GNU 9.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting default Kokkos CXX standard to 17
-- Kokkos version: 4.1.00
-- The project name is: Kokkos
-- Using internal gtest for testing
-- Configured git information in /home/jacob/projects/experiments/scratch/build/generated/Kokkos_Version_Info.cpp
-- Compiler Version: 11.7.64
-- kokkos_launch_compiler (/home/jacob/projects/experiments/scratch/kokkos/bin/kokkos_launch_compiler) is enabled...
-- SERIAL backend is being turned on to ensure there is at least one Host space. To change this, you must enable another host execution space and configure with -DKokkos_ENABLE_SERIAL=OFF or change CMakeCache.txt
-- Using -std=c++17 for C++17 standard as feature
-- Built-in Execution Spaces:
--     Device Parallel: Kokkos::Cuda
--     Host Parallel: NoTypeDefined
--       Host Serial: SERIAL
--
-- Architectures:
--  ZEN2
--  AMPERE86
-- Found CUDAToolkit: /usr/local/cuda-11.7/include (found version "11.7.64")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found TPLCUDA: TRUE
-- Found TPLLIBDL: /usr/include
-- Using internal desul_atomics copy
-- Kokkos Devices: CUDA;SERIAL, Kokkos Backends: CUDA;SERIAL
-- Configuring done
-- Generating done
-- Build files have been written to: /home/jacob/projects/experiments/scratch/build
```

6. code needed to reproduce the bug

Here is a minimal reproducible example I constructed:
```c++
#include <Kokkos_Core.hpp>

int main(int argc, char *argv[]) {
  Kokkos::initialize(argc, argv);
  {
  
  int scr_level = 1;
  int scr_size = 4896;

  const int nk = 7535;

  Kokkos::TeamPolicy<> policy(Kokkos::DefaultExecutionSpace(), nk, Kokkos::AUTO);
  policy.set_scratch_size(scr_level, Kokkos::PerTeam(scr_size));
  Kokkos::parallel_for("kernel", policy,
  KOKKOS_LAMBDA(Kokkos::TeamPolicy<>::member_type tmember) {
  });

  }
  Kokkos::finalize();

  return 0;
}
```

Here's the CMakeLists.txt file:
[CMakeLists.txt](https://github.com/kokkos/kokkos/files/12477257/CMakeLists.txt)

The CMake configuration expects a local copy of Kokkos rather than a globally installed one.

8. command line needed to reproduce the bug
```
make
./scratch
```
10. please also attach the `KokkosCore_config.h` header file (generated during the build);
[KokkosCore_config.zip](https://github.com/kokkos/kokkos/files/12477401/KokkosCore_config.zip)

**Any additional info**
Setting `scr_level = 0` fixes the issue, but unfortunately my intended application needs level 1 scratch. I have observed this problem on both my local machine and Perlmutter, which use Ampere-class GPUs. A separate machine using a Volta-class GPU seems to run just fine.

Here is an abbreviated device stack trace from `compute-sanitizer`:
```
========= COMPUTE-SANITIZER
========= Invalid __global__ atomic of size 4 bytes
=========     at 0x5c0 in /usr/local/cuda-11.7/include/device_atomic_functions.hpp:172:_INTERNAL_cf6669a5_8_main_cpp_main::atomicCAS(int *, int, int)
=========     by thread (0,0,0) in block (489,0,0)
=========     Address 0x608b0f600 is out of bounds
=========     and is 4609 bytes after the nearest allocation at 0x608ad2400 of size 245760 bytes
=========     Device Frame:/home/jacob/projects/experiments/scratch/kokkos/core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp:419:Kokkos::Impl::cuda_get_scratch_index(unsigned int, int *, unsigned long) [0x510]
=========     Device Frame:/home/jacob/projects/experiments/scratch/kokkos/core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp:497:Kokkos::Impl::ParallelFor<main::[lambda(Kokkos::Impl::CudaTeamMember) (instance 1)], Kokkos::TeamPolicy<>, Kokkos::Cuda>::operator ()() const [0x50]
=========     Device Frame:/home/jacob/projects/experiments/scratch/kokkos/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:88:void Kokkos::Impl::cuda_parallel_launch_local_memory<Kokkos::Impl::ParallelFor<main::[lambda(Kokkos::Impl::CudaTeamMember) (instance 1)], Kokkos::TeamPolicy<>, Kokkos::Cuda>>(T1) [0x10]
=========     Saved host backtrace up to driver entry point at kernel launch time
```

On further investigation, I wonder if the private variable `m_num_scratch_locks` is actually uninitialized for `ParallelFor` in the CUDA implementation. Doing something analogous to `ParallelReduce` and adding `m_num_scratch_locks = internal_space_instance->m_num_scratch_locks;` after line 559 in `Kokkos_Cuda_Parallel_Team.hpp` seems to fix the issue. With that change, my toy problem and the code I originally found the problem in both seem to work, and `compute-sanitizer` comes back clean now. That being said, I'm unfamiliar with the Kokkos codebase, so I don't know if this breaks other parts of the code.

**Comment by ajpowelsnl** [comment]
Hi @jfields7 - I can give it a try on Perlmutter, and see if we can cobble together a fix.  Many thanks for reporting this issue.

**Comment by fnrizzi** [comment]
relative PR was merged, so closing this for now

**Comment by masterleinad** [comment]
Let's do our usual thing and only close after the release (or when the issue was only present in `develop` anyway.

**Comment by fnrizzi** [comment]
ok

**Comment by masterleinad** [comment]
Fixed in #6433.
